mpf/devices/multiball.py

```diff
@@ -291,6 +291,54 @@ def add_a_ball(self):
             self.balls_added_live += 1
             self.machine.game.balls_in_play += 1
             self.source_playfield.add_ball(balls=1)
+            self.add_a_ball_timer_start()
+
+    def add_a_ball_timer_start(self) -> None:
+        """Start the timer for add a ball ball save.
+
+        This is started when multiball add a ball is triggered if configured,
+        and the default timer is not still running.
+        """
+        if self.shoot_again:
+            # if main ball save timer is running, don't run this timer
+            return
+        self.shoot_again = True
+
+        shoot_again_ms = self.config['add_a_ball_shoot_again'].evaluate([])
+        if not shoot_again_ms:
+            # No shoot again. Just stop multiball right away
+            self.stop()
+        else:
+            # Enable shoot again
+            self.machine.events.add_handler('ball_drain',
+                                            self._ball_drain_shoot_again,
+                                            priority=1000)
+
+        self.machine.events.post('ball_save_{}_add_a_ball_timer_start'.format(self.name))
+        '''event: ball_save_(name)_add_a_ball_timer_start
+        desc: The multiball add a ball ball save called (name) has just start its countdown timer.
+        '''
+
+        grace_period_ms = self.config['add_a_ball_grace_period'].evaluate([])
+        hurry_up_time_ms = self.config['add_a_ball_hurry_up_time'].evaluate([])
+        if shoot_again_ms > 0:
```

Comment:
If those are the same handlers as in the normal shoot again case you could refactor this into a separate method with takes `grace_period_ms` and  `hurry_up_time_ms` as methods.

Reply:
Not 100% sure I understand how to execute the ask here.  They are different config section.  You have grace_period for normal ball save and add_a_ball_grace_period for the add a ball (so that they can be different values).  We then need to calculate them, and add handlers if needed.

If that means we are good as is, then I will get the other stuff pushed.  If not, I think you are meaning create a method that takes a config section and performs the conversion and returns the ms.  So instead of doing this inline, we would call the method to do it?

Reply:
I propose a method such as:

```
    def _start_shoot_again(self, shoot_again_ms, hurry_up_time_ms):
```
This would then just contain the line below this comment and be called `self._start_shoot_again(shoot_again_ms, hurry_up_time_ms)` from there and from `_timer_start`. Does that make sense?

Reply:
Got it, that makes sense.  Should be there now where both places can call the one method, and they evaluate the variables first from the config, but then pass them in for handlers to be setup.  Good idea to reuse code and make it easier to maintain going forward.